pkg/agent/controller/networkpolicy/audit_logging.go

```diff
@@ -0,0 +1,152 @@
+func (a *AntreaPolicyLogger) terminateLogKey(logMsg string) {
+	a.logDeduplication.logMutex.Lock()
+	defer a.logDeduplication.logMutex.Unlock()
+	if a.logDeduplication.logMap[logMsg].count == 1 {
+		a.anpLogger.Printf(logMsg)
+	} else {
+		a.anpLogger.Printf("%s [%d packets in %s]", logMsg, a.logDeduplication.logMap[logMsg].count, time.Since(a.logDeduplication.logMap[logMsg].initTime))
+	}
+	delete(a.logDeduplication.logMap, logMsg)
+}
+
+// updateLogKey initiates record or increases the count in logDeduplication corresponding to given logMsg.
+func (a *AntreaPolicyLogger) updateLogKey(logMsg string, bufferLength time.Duration) bool {
+	a.logDeduplication.logMutex.Lock()
+	defer a.logDeduplication.logMutex.Unlock()
+	_, exists := a.logDeduplication.logMap[logMsg]
+	if exists {
+		a.logDeduplication.logMap[logMsg].count++
+	} else {
+		record := logDedupRecord{1, time.Now(), time.NewTimer(bufferLength)}
+		a.logDeduplication.logMap[logMsg] = &record
+	}
+	return exists
+}
+
+// logDedupPacket logs information in ob based on disposition and duplication conditions.
+func (a *AntreaPolicyLogger) logDedupPacket(ob *logInfo, bufferLength time.Duration) {
+	// Deduplicate non-Allow packet log.
+	logMsg := fmt.Sprintf("%s %s %s %s SRC: %s DEST: %s %d %s", ob.tableName, ob.npRef, ob.disposition, ob.ofPriority, ob.srcIP, ob.destIP, ob.pktLength, ob.protocolStr)
+	if ob.disposition == openflow.DispositionToString[openflow.DispositionAllow] {
+		a.anpLogger.Printf(logMsg)
+	} else {
+		// Increase count if duplicated within 1 sec, create buffer otherwise.
+		exists := a.updateLogKey(logMsg, bufferLength)
+		if !exists {
+			// Go routine for logging when buffer timer stops.
+			go a.logAfterTimer(logMsg)
+		}
+	}
+}
+
+// newAntreaPolicyLogger is called while newing Antrea network policy agent controller.
+// Customize AntreaPolicyLogger specifically for Antrea Policies audit logging.
+func newAntreaPolicyLogger() (*AntreaPolicyLogger, error) {
+	logDir := filepath.Join(logdir.GetLogDir(), logfileSubdir)
+	logFile := filepath.Join(logDir, logfileName)
+	_, err := os.Stat(logDir)
+	if os.IsNotExist(err) {
+		os.Mkdir(logDir, 0755)
+	} else if err != nil {
+		return nil, fmt.Errorf("received error while accessing Antrea network policy log directory: %v", err)
+	}
+
+	// Use lumberjack log file rotation.
+	logOutput := &lumberjack.Logger{
+		Filename:   logFile,
+		MaxSize:    500,  // allow max 500 megabytes for one log file
+		MaxBackups: 3,    // allow max 3 old log file backups
+		MaxAge:     28,   // allow max 28 days maintenance of old log files
+		Compress:   true, // compress the old log files for backup
+	}
+
+	cAntreaPolicyLogger := &AntreaPolicyLogger{
+		anpLogger:        log.New(logOutput, "", log.Ldate|log.Lmicroseconds),
+		logDeduplication: logRecordDedupMap{logMap: make(map[string]*logDedupRecord)},
+	}
+	klog.V(2).Infof("Initialized Antrea-native Policy Logger for audit logging with log file '%s'", logFile)
```

Comment:
If you are going to make a change, better to change this logging to structured logging too.